### ISSUE TITLE
Upgrade Chevrotain to 6.x

### ISSUE THREAD

**Issue by bd82** [jhipster/prettier-java#248]
There have been quite a few changes including breaking ones.
https://sap.github.io/chevrotain/docs/changes/CHANGELOG.html#_6-1-0-8-25-2019

**Comment by clement26695** [comment]
I'll try to do the update

**Comment by bd82** [comment]
You will probably need to wrap the custom backtracking logic with `$.ACTION` 
- https://sap.github.io/chevrotain/docs/guide/internals.html#grammar-recording

**Comment by bd82** [comment]
Version 6.3 includes features and a guide related to improving initialization performance.
- https://sap.github.io/chevrotain/docs/guide/initialization_performance.html

**Comment by Shaolans** [comment]
With @clement26695, we are looking to update prettier-java to the latest version of Chevrotain.
However we do not completely understand how to use ACTION in our backtracking logic since they are using rules.
Would it be possible to show us how to refactor this logic with the example below ?

```javascript
  $.RULE("isCastExpression", () => {
    if (this.BACKTRACK_LOOKAHEAD($.isPrimitiveCastExpression)) {
      return true;
    }
    return this.BACKTRACK_LOOKAHEAD($.isReferenceTypeCastExpression);
  });

  $.RULE("isPrimitiveCastExpression", () => {
    $.CONSUME(t.LBrace);
    $.SUBRULE($.primitiveType);
    // No dims so this is not a reference Type
    $.CONSUME(t.RBrace);
    return true;
  });
```

**Comment by bd82** [comment]
Try wrapping the whole `isCastExpression` in "ACTION".
For `isPrimitiveCastExpression` you should not need to do anything.

```javascript
  $.RULE("isCastExpression", () => {
    return $.ACTION(() => {
           if (this.BACKTRACK_LOOKAHEAD($.isPrimitiveCastExpression)) {
        return true;
      }
      return this.BACKTRACK_LOOKAHEAD($.isReferenceTypeCastExpression);
      })
  });
```

**Comment by bd82** [comment]
Basically you want to avoid running backtracking logic during the grammar recording phase

**Comment by clement26695** [comment]
I have to admit that I am a bit lost here... I just tried to upgrade it and wrapped every Backtracking logic in an ACTION (see [here](https://github.com/clement26695/prettier-java/pull/4)), but I got an error when I try to parse any file, for instance this one:

```java
public class App {

  private static final Logger LOGGER = LoggerFactory.getLogger(App.class);

  public App() {
    LOGGER.info("Constructing parts and car");
  }
}
```

I got this stacktrace: 
```
TypeError: Cannot read property 'call' of undefined
    at JavaParser.RecognizerEngine.optionInternalLogic (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:227:27)
    at JavaParser.RecognizerEngine.optionInternal (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:196:25)
    at JavaParser.RecognizerApi.OPTION (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_api.js:102:21)
    at JavaParser.$.ACTION (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/productions/packages-and-modules.js:215:9)
    at JavaParser.RecognizerApi.ACTION (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_api.js:21:21)
    at JavaParser.$.RULE (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/productions/packages-and-modules.js:214:14)
    at JavaParser.invokeRuleWithTry (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:119:33)
    at JavaParser.wrappedGrammarRule [as isModuleCompilationUnit] (/home/cdessoude/Documents/workplace/prettier-java/node_modules/chevrotain/lib/src/parse/parser/traits/recognizer_engine.js:133:38)
    at JavaParser.BACKTRACK_LOOKAHEAD (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/parser.js:183:25)
    at JavaParser.$.RULE (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/productions/packages-and-modules.js:8:24)
```

**Comment by bd82** [comment]
- Never wrap DSL methods (CONSUME/OPTION/SUBRULE/...) in ACTION.
- Only wrap the invocation to the backtracking utilities in ACTION.

Bad example: https://github.com/clement26695/prettier-java/pull/4/files#diff-3be86ec11b8168a43f3526a40d5af78eR214

Basically You want the "recording phase" to run on your rules to build the data structure representing the grammar. so it cannot be wrapped in ACTION as this skips its invocation during the recording phase. However you do not want to RECORD the backtracking logic as that is not part of the grammar, so you wrap calls to `isXYZ` in ACTION.

btw: GATES are not executed during backtracking so no need to wrap code already inside GATE property with ACTION.

This is a pretty complex grammar in terms of using advanced optimized backtracking... 
If you are still stuck I will try to convert myself, perhaps on the weekend...

**Comment by clement26695** [comment]
Thank you for the clarifications ! Will try to upgrade this week, I will keep you posted

**Comment by clement26695** [comment]
Closed with #268
